src/main/java/net/minecraftforge/event/world/BlockEvent.java

```diff
@@ -415,65 +417,153 @@ public PortalShape getPortalSize()
     @Cancelable
-    public static class BlockToolInteractEvent extends BlockEvent
+    @Deprecated(forRemoval = true, since = "1.18.2")
+    public static class BlockToolInteractEvent extends BlockToolModificationEvent
     {
-
         private final Player player;
         private final ItemStack stack;
-        private final ToolAction toolAction;
-        private BlockState state;
+
+        public BlockToolInteractEvent(BlockState originalState, @Nonnull UseOnContext context, ToolAction toolAction)
+        {
+            super(originalState, context, toolAction, false);
+            this.player = context.getPlayer();
+            this.stack = context.getItemInHand();
+        }
 
         public BlockToolInteractEvent(LevelAccessor world, BlockPos pos, BlockState originalState, Player player, ItemStack stack, ToolAction toolAction)
         {
-            super(world, pos, originalState);
+            super(world, pos, originalState, toolAction);
             this.player = player;
             this.stack = stack;
+        }
+
+        public Player getPlayer()
+        {
+            return this.player;
+        }
+
+        public ItemStack getHeldItemStack()
+        {
+            return this.stack;
+        }
+    }
+
+    /**
+     * Fired when a block is right-clicked by a tool to change its state.
+     * For example: Used to determine if {@link ToolActions#AXE_STRIP an axe can strip},
+     * {@link ToolActions#SHOVEL_FLATTEN a shovel can path}, or {@link ToolActions#HOE_TILL a hoe can till}.
+     * <p>
+     * Care must be taken to ensure world-modifying events are only performed if {@link #isSimulated()} returns {@code true}.
+     * <p>
+     * This event is {@link Cancelable}. If canceled, this will prevent the tool
+     * from changing the block's state.
+     */
+    public static class BlockToolModificationEvent extends BlockEvent
+    {
+        private final UseOnContext context;
+        private final ToolAction toolAction;
+        private final boolean simulate;
+        private BlockState state;
+
+        public BlockToolModificationEvent(BlockState originalState, @Nonnull UseOnContext context, ToolAction toolAction, boolean simulate)
+        {
+            super(context.getLevel(), context.getClickedPos(), originalState);
+            this.context = context;
             this.state = originalState;
             this.toolAction = toolAction;
+            this.simulate = simulate;
         }
 
-        /**Gets the player using the tool.*/
+        // TODO 1.19: Remove
+        BlockToolModificationEvent(LevelAccessor level, BlockPos pos, BlockState originalState, ToolAction toolAction)
+        {
+            super(level, pos, originalState);
+            this.context = null;
+            this.state = originalState;
+            this.toolAction = toolAction;
+            this.simulate = false;
+        }
+
+        /**
+         * @return the player using the tool
+         */
         public Player getPlayer()
```

Comment:
I feel like this should be marked as nullable and then just marked as nonnull in the override? Mostly given the fact that `Player` in `UseOnContext` is nullable